### DISCUSSION TITLE
Invoice payment

### DISCUSSION THREAD

**Discussion by JohnG-cc** [CredComSoc/ProtocolWorkingGroup#10] [Track 2 Ideas, requirements and features]
Moving a discussion across from Telegram:

@JohnG-cc wrote:
> A possibly related question; if someone several nodes over issues me an invoice in their currency, and there are multiple currency conversions and rounding errors between me and them, is there a way I can determine how much of my currency to send so that they receive the exact amount that they are expecting?

@matslats wrote:
> The exact amount, no. It could be one millionth out per node

@JohnG-cc wrote:
> I guess that’s inevitable with any exchange rate even without rounding errors. Is there a way to determine an approximate “at least enough to settle the invoice” amount?

@matslats wrote:
> good question.

As far as I can tell, this is not yet in the protocol(?) and probably needs to be (??)

Reflecting on this though, it seems like the exchange rates are more of an issue in this context because I'm trying to do a one-shot push transfer, and hoping the right amount arrives on the other end. It seems like this might be a better user experience if the protocol had the facility to allow source and destination nodes to negotiate the amount to be sent, so that client apps could present the proposed source and dest amounts to the user for confirmation, and then also confirm that the proposed minimum destination amount has been received (???)

**Comment by dilgreen** [comment]
When we check that a proposed transaction is achievable (is acceptable to all nodes on the path), we adopt a 'lease' method. Each node guarantees a duration during which it will honour its declaration that a transaction is acceptable.

This means that, even if balance limits are exceeded due to other transactions being completed during the lease time, the node will process it.

I am wondering if a similar approach to exchange rates might help?

**Reply by dilgreen** [reply]
Further, I wonder if nodes might not maintain (and periodically update) a cached picture of their path to the trunk, including member lists of all nodes en-route, lease times and exchange rates?

Then a proposed transaction can be quickly assessed with two calls - one to the proposing node (ie internal), which uses its cache to determine the conditions to deepest branch, and one to the receiving node, providing the other half of the picture.

This cache could be maintained very cheaply through the members of the first tier of branches polling the trunk (which will be optimised for event driven work), then a recursive process whereby each node informs its members if there is any change to the cached data, or it it has changed its own data.

**Reply by dilgreen** [reply]
Further, if each node asks its members what their current lease time is  , then the first tier of branches can set their polling frequency to that of the most time-sensitive node in their partition of the tree.

I think this would mean that accurate and stable predictions of price and guaranteed stability could be delivered more cheaply, speedily, and robustly, without violating the data sovereignty of leaf nodes, or granting too much information asymmetry trunkward (the only trunkward data aggregation is greater oversight of where the nodes are in the network which are currently granting the shortest leases. This might be a bad idea).

**Reply by dilgreen** [reply]
There is a whole instability issue arising from all this business of transaction conditions at each node being subject to variance - which is the behaviour of the system in crisis.

On the one hand, imposing minimum intervals between policy changes increases stability (and is significantly better for markets in normal times), but decreases ability of nodes to respond to events.

On the other hand, allowing eg algorithmic / market driven changes to policy increases sensitivity to events, but decreases stability.

My instinct is to go with:

1. an interval limit system (a policy) governing how often lease times and exchange rates may be altered, which is, in practice, set as a default that is  at some guesstimate of a sweet spot, erring on the side of stability. 
2. no machinery which allows programmatic changes to these interval settings (although a way to do this if considered important)
3. machinery which allows programmatic changes to leases, exchange rates an other 'direct' policies (frequency limited by the interval policy)

**Comment by JohnG-cc** [comment]
@dilgreen wrote:
> then a recursive process whereby each node informs its members if there is any change to the cached data, or it it has changed its own data.

That's an interesting idea. So each node would have a theoretically up-to-date map of trunkward conditions. That would cut down on a lot of chatter. It might also allow payer and payee nodes to define invoice amounts in the common currency of the trunk node, and just convert to node-local currency for display. An invoice denominated in trunk currency could be sent directly node to node over the internet, or could be routed across the tree with and "invoice" message type (the latter seems more parsimonious)

**Comment by JohnG-cc** [comment]
Trunk denominated invoices also splits exchange rate risk down the middle - each party is only exposed to rate movements on "their side" of the tree during the term of the invoice.

This sort of model suggests making invoices a persistent entity in a node. This is a feature that for example paypal provides.

**Comment by JohnG-cc** [comment]
Of course an invoice could also be denominated in payer or payee node currency if the parties saw fit to manage exchange rate risk that way.

**Comment by dilgreen** [comment]
It may be worth unpacking the meaning of the term 'invoice' in a mutual credit context.

In Mutual Credit systems,  a variety of workflows for transactions may be implemented. Afaik, CCP/CredCom is relatively un-opinionated about workflow specifics - allowing group-specific policies to be applied.

The most 'full fat' Mutual Credit workflow specification would look like this;

1. Either party to initiate a transaction - ie an 'invoice'/request for payment from a payee _or_ a 'transfer-to' instruction request to pay from a payer.
2. Provide the counterparty with information about the pending transaction, including any fees
3. Require a positive acceptance from the counterparty
4. Expect all parties to immediately adjust their balances to reflect the transaction.

Variants might include:

- requiring all transactions to be initiated by payer (or payee)
- not requiring positive acceptance (also, allowing users/groups to set a policy (presumably parameterised) for auto-acceptance by a user agent (client software)
- Allowing for delayed transaction completion

_Note_: relating to point 3 above, something I read suggested that it might be potentially helpful from a legal point of view to implement this 'acceptance' requirement (which would not stop it from being programmatically implemented). This usefulness would stem from it being possible to present the transaction itself as a contract (as separate from any contract to supply which might be the context for the transaction). A typical bank payment is directed entirely by the payer, with the payee's bank passively accepting on behalf of the unknowing payee. This arrangement has of course been the subject of law and jurisprudence worldwide. Mutual Credit has not. Being able to present the transaction itself as a contract, it has been suggested to me (although not as legal advice), might be a protective measure against legal/regulatory attack - since contracts are generally sanctified in law and jurisprudence.

**Reply by dilgreen** [reply]
The above suggests to me that we might need to get a parallel working group - perhaps through the Credit Commons Society - to set out a description of the social/legal arrangements of  Mutual Credit which this Protocol needs to serve / enable.

They are implicit, but not explicit in the workings of CCP/CredCom.

@matslats I am sure this sort of thing must exist in various places, from different authors, but it seems dangerous not to have a version for use as reference here.

**Reply by bu6n** [reply]
There's a slight difference I think, because most bank payments do not have fees. In the context of the credit commons, if the payee also has to pay fees, then he probably wants to sign the transaction first, otherwise an adversary node could setup ridiculous fees and extract credits from other parties. Maybe there should be a way for a user to allow incoming transactions with a limit on fees, but even then, nothing would stop someone a node in-between to randomly augment fees to be just under the limit...

**Comment by matslats** [comment]
The workflows defined in the protocol allow a confirmation stage where you can see the built transaction before confirming it AND or the other party can approve it before it is completed too.
The system isn't built for adversaries, which is to say that membership implies being under an umbrella of trust. 
I'm not sure how this doesn't meet the requirements described above.

**Reply by bu6n** [reply]
I'm not sure if this is a response to my comment. I was responding to @dilgreen about this mainly: "A typical bank payment is directed entirely by the payer, with the payee's bank passively accepting on behalf of the unknowing payee. This arrangement has of course been the subject of law and jurisprudence worldwide." I just meant it's not easy to do automatic payments if every node in the middle can have fees. Doesn't to be an adversary, could also just be a security breach or whatever.
